### PR TITLE
Prevent curators from creating new instances of species, compartments.

### PR DESCRIPTION
**Pull request by lbeckman314**
- Checks for "uneditable" classes in schema view.
- Adds event listener for edit menu to enable/disable based on class.

Let me know if there is anything at all that you think could be added, removed, or updated in this pull request to better suit it's adoption! : )

### CODE REVIEW

File: src/org/gk/gkCurator/CuratorActionCollection.java [hunk 1]
```diff
@@ -2169,10 +2169,41 @@ private void deleteInstance() {
...
 	
+	/**
+	 * Retrieve a "Create Instance" action for new entities.
+	 * 
+	 * @return Action with specific "SHORT_DESCRIPTION" and "enabled" status
+	 */
 	public Action getCreateInstanceAction() {
-		if (createInstanceAction == null) {
+		// list of constants that curators are not allowed to create new instances of.
+		ArrayList<String> nonCreatable = new ArrayList<String>();
```
Comment (jweiser) [inline src/org/gk/gkCurator/CuratorActionCollection.java]: Could the declaration be for `List<String>`?

File: src/org/gk/gkCurator/CuratorActionCollection.java [hunk 2]
```diff
@@ -2169,10 +2169,41 @@ private void deleteInstance() {
...
+		nonCreatable.add(ReactomeJavaConstants.Species);
+		nonCreatable.add(ReactomeJavaConstants.Compartment);
```
Comment (jweiser) [inline src/org/gk/gkCurator/CuratorActionCollection.java]: If `nonCreatable` will be a fixed size, an alternative initialization could be:

```
List<String> nonCreatable = Arrays.asList(
    ReactomeJavaConstants.Species, 
    ReactomeJavaConstants.Compartment
);
```

File: src/org/gk/gkCurator/CuratorActionCollection.java [hunk 3]
```diff
@@ -2169,10 +2169,41 @@ private void deleteInstance() {
...
+		GKSchemaClass selectedClass = null;
+		if (curatorFrame.getFocusedComponent() instanceof SchemaViewPane) {
+			SchemaViewPane schemaViewPane = (SchemaViewPane) curatorFrame.getFocusedComponent();
+			selectedClass = schemaViewPane.getSchemaPane().getSelectedClass();
+		}
+
+		// if creation of instance is not allowed, then user is notified to contact editor.
+		if (selectedClass != null && nonCreatable.contains(selectedClass.getName())) {
+			// This is an information for blocking creating a new species and compartment:
+			String message = "You cannot create an instance for this type of class. Please ask Peter D’Eustachio to create one for you.";
```
Comment (jweiser) [inline src/org/gk/gkCurator/CuratorActionCollection.java]: Maybe the person to contact could be made a constant in this class or another class managing constant values?

File: src/org/gk/gkCurator/CuratorActionCollection.java (hunk 1 above)
Comment (lbeckman314) [reply src/org/gk/gkCurator/CuratorActionCollection.java]: Absolutely! I will update.

File: src/org/gk/gkCurator/CuratorActionCollection.java (hunk 2 above)
Comment (lbeckman314) [reply src/org/gk/gkCurator/CuratorActionCollection.java]:
```java
List<String> nonCreatable = Arrays.asList(
    ReactomeJavaConstants.Species, 
    ReactomeJavaConstants.Compartment
);
```
This is much better approach!

File: src/org/gk/gkCurator/CuratorActionCollection.java (hunk 3 above)
Comment (lbeckman314) [reply src/org/gk/gkCurator/CuratorActionCollection.java]: I will migrate the name over to a constant value class!

Do you think the uneditable classes (`species` and `compartment` in this example), be in a separate class as well to keep things easier to maintain?

Comment (guanmingwu) [reply src/org/gk/gkCurator/CuratorActionCollection.java]: Liam, As we discussed, these values (un-editable classes) should be configured inside the file resources/curator.xml. If you like, you can also configure the contact name.

**Comment by lbeckman314** [comment]
Oops -- I accidentally clicked too fast and hit "resolved" on the last review. Is there a way to bring it back from the grave?

---

I just talked to Guanming and I got some great pointers:
- I can revert the above changes and simply use the existing `resources/curator.xml` file to update uneditable classes. Makes things a lot especially for Peter and other editors!
- In order to keep things simple, I'll branch off the `reactome:CuratorTool`'s `master` branch, instead of off my own `lbeckman314:CuratorTool` branches.

Thank you for reviewing everything so fast!

File: src/org/gk/gkCurator/CuratorActionCollection.java (hunk 1 above)
Comment (jweiser) [reply src/org/gk/gkCurator/CuratorActionCollection.java]: Great, thanks!

**Comment by jweiser** [comment]
> Oops -- I accidentally clicked too fast and hit "resolved" on the last review. Is there a way to bring it back from the grave?
> 
> I just talked to Guanming and got some great pointers:
> 
> * I can revert the above changes and simply use the existing `resources/curator.xml` file to update uneditable classes. Makes things a lot especially for Peter and other editors!
> * In order to keep things simple, I'll branch off the `reactome:CuratorTool`'s `master` branch, instead of off my own `lbeckman314:CuratorTool` branches.
> 
> Thank you for reviewing everything so fast!

Yup, if you have a resolved conversation in this tab of the pull request, you can expand it by clicking "show resolved" at the right and then clicking "Unresolve conversation"

**Comment by lbeckman314** [comment]
Got the new pull request up and running -- thank you very much for your help!

https://github.com/reactome/CuratorTool/compare/prevent-species-creation?expand=1

**Pull request closed without merging**